usdigitalresponse/grants-ingest
# 61: Receive new FFIS digest emals

### Ticket #61 

Closes #61 

## Description

This PR adds a (previously-missing) Lambda handler that verifies the sender (and general trustworthiness) of emails delivered by SES to the email delivery bucket. Additionally, it modifies the terraform related the `PersistFFISData` Lambda function, which was previously misconfigured in terms of which bucket was expected to trigger its invocations (for the record, `Persist*` Lambda functions are invoked by the _prepared_ data bucket, not the raw source data bucket). A few other terraform modifications are contained as well, but these are minor tweaks made for the sake of housekeeping and general consistency.

Summary of changes:
- New Lambda runtime code at `cmd/ReceiveFFISEmail`:
  - This function is invoked when emails are delivered to the bucket we've dedicated for SES email delivery. 
  - Upon invocation, it inspects the new email to verify that the sender is allow-listed. Additionally, it ensures that SES spam and virus checks were marked as successful. 
  - Once these verifications are complete and passing, the Lambda copies the received email to the source data bucket for further processing (specifically, the next step is the `EnqueueFFISDownload` function).
- New terraform module corresponding to the new Lambda function at `terraform/modules/ReceiveFFISEmail`. This is all pretty standard/similar to the other Lambda-specific terraform modules.
- Fixes in Terraform to the `PersistFFISData` Lambda function's invocation configuration. Previously, it was configured to be invoked by the wrong bucket.
- Other miscellany:
  - Update `terraform/outputs.tf` to include the names of all Lambda functions in output. Previously a few were missing.
  - Simple refactor (across terraform and runtime code) to make S3 key paths consistent for FFIS-related objects. Previously, there was a mix of `/ffis/` and `/ffis.org/` path segments, which was leading to some inconsistencies in how/whether a Lambda would be invoked in response to a new bucket object. It was also just generally confusing to keep track of.

## Testing

1. Apply terraform to a LocalStack environment, potentially configuring the `ffis_email_allowed_senders` input variable to include allow-listed email addresses and/or domain names of your choosing.
2. Use any of the fixture `.eml` (email) files in `cmd/ReceiveFFISEmail/fixtures/` as a starting point, and upload them to the email delivery bucket in the LocalStack environment (see the `module.email_delivery_bucket` resource) with a key that begins with `ses/ffis_ingest/new/`.
3. For good emails, the resulting `ReceiveFFISEmail` Lambda invocation should succeed and copy the file to the source data bucket (see the `module.grants_source_data_bucket` resource) at a path of `sources/yyyy/mm/dd/ffis.org/raw.eml`, where the date-based components of the path correspond to the value of the `Date` header in the email file. For bad emails, the invocation should fail without copying the S3 object.
    - "Good" emails means an email file that is valid/parseable, where the `From` header contains an email address that is allow-listed (the allow-list that you set in the `ffis_email_allowed_senders` can contain a list of domain names and/or full email addresses), an `X-SES-SPAM-VERDICT: PASS` header, an `X-SES-VIRUS-VERDICT: PASS` header, and a `Received-SPF` header that begins with the word `pass`. 
    - "Bad" emails are anything else :)


### Automated and Unit Tests

### Manual tests for Reviewer

## Checklist
